src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+         int inputBlocksOffset {0};
+
+         data.inputs = static_cast<Vst::AudioBusBuffers*>(
+            alloca(sizeof(Vst::AudioBusBuffers) * data.numInputs));
+
+         for(int busIndex = 0; busIndex < data.numInputs; ++busIndex)
+         {
+            Vst::BusInfo busInfo;
+            if(effect->getBusInfo(Vst::kAudio, Vst::kInput, busIndex, busInfo) != kResultOk)
+            {
+               wxLogError("Failed to read bus info");
+               return 0;
+            }
+            if(busInfo.busType == Vst::kMain)
+            {
+               data.inputs[busIndex].numChannels = busInfo.channelCount;
+               data.inputs[busIndex].channelBuffers32 = inBlock + inputBlocksOffset;
+               inputBlocksOffset += busInfo.channelCount;
+            }
+            else
+            {
+               //aux is not yet supported
+               data.inputs[busIndex].numChannels = 0;
+               data.inputs[busIndex].channelBuffers32 = nullptr;
+            }
+            data.inputs[busIndex].silenceFlags = 0UL;
+         }
+      }
+      if(data.numOutputs > 0)
+      {
+         int outputBlocksOffset {0};
+
+         data.outputs = static_cast<Vst::AudioBusBuffers*>(
+            alloca(sizeof(Vst::AudioBusBuffers) * data.numOutputs));
+         for(int busIndex = 0; busIndex < data.numOutputs; ++busIndex)
+         {
+            Vst::BusInfo busInfo;
+            if(effect->getBusInfo(Vst::kAudio, Vst::kOutput, busIndex, busInfo) != kResultOk)
+            {
+               wxLogError("Failed to read bus info");
+               return 0;
+            }
+            if(busInfo.busType == Vst::kMain)
+            {
+               data.outputs[busIndex].numChannels = busInfo.channelCount;
+               data.outputs[busIndex].channelBuffers32 = outBlock + outputBlocksOffset;
+               outputBlocksOffset += busInfo.channelCount;
+            }
+            else
+            {
+               //aux is not yet supported
+               data.outputs[busIndex].numChannels = 0;
+               data.outputs[busIndex].channelBuffers32 = nullptr;
+            }
+            data.outputs[busIndex].silenceFlags = 0UL;
+         }
+      }
+      
+      const auto processResult = audioProcessor->process(data);
+      
+      return processResult == kResultOk ?
+         data.numSamples : 0;
+   }
+   wxLogError("Failed to query IAudioProcess interface from the effect component");
+   return 0;
+   }
+}
+
+size_t VST3Effect::ProcessBlock(float** inBlock, float** outBlock, size_t blockLen)
+{
+   internal::ComponentHandler::PendingChangesPtr pendingChanges { nullptr };
+   if(mComponentHandler)
+      pendingChanges = mComponentHandler->getPendingChanges();
+   return VST3ProcessBlock(mEffectComponent.get(), mSetup, inBlock, outBlock, blockLen, pendingChanges.get());
+}
+
+bool VST3Effect::RealtimeInitialize()
+{
+   //reload current parameters form the editor into parameter queues
+   SyncParameters();
+   return true;
```

Comment:
For the VST2 effect, ProcessInitialize is called.  For VST3, you only `SyncParameters` but don't do the other side effects.  Is it correct?

Reply:
I believe so, the `RealtimeInitialize` is called before any call to `RealtimeAddProcessor`, so we cannot configure them here (but they are initialized in `RealtimeAddProcessor` instead), and there is no need to set up the "container" effect since it's not used for realtime processing.